tests/test_dragon_installer.py

```diff
@@ -102,23 +103,25 @@ def test_assets(monkeypatch: pytest.MonkeyPatch) -> t.Dict[str, GitReleaseAsset]
     assets: t.List[GitReleaseAsset] = []
     mock_archive_name_tpl = "{}-{}.4.1-{}ac132fe95.tar.gz"
 
-    for python_version in ["py3.9", "py3.10", "py3.11"]:
-        for dragon_version in ["dragon-0.8", "dragon-0.9", "dragon-0.10"]:
-            for platform in ["", "CRAYEX-"]:
-
-                asset = GitReleaseAsset(requester, headers, attributes, completed)
-
-                archive_name = mock_archive_name_tpl.format(
-                    dragon_version, python_version, platform
-                )
-
-                monkeypatch.setattr(
-                    asset,
-                    "_browser_download_url",
-                    _git_attr(value=f"http://foo/{archive_name}"),
-                )
-                monkeypatch.setattr(asset, "_name", _git_attr(value=archive_name))
-                assets.append(asset)
+    for python_version, dragon_version, platform in itertools.chain(
+        itertools.product(
+            ["py3.9", "py3.10", "py3.11"], ["dragon-0.8", "dragon-0.9"], ["", "CRAYEX-"]
```

Comment:
Oh! Was the point of this test not mimic the names of the assets from the [dragon releases](https://github.com/DragonHPC/dragon/releases)?

If so, I'm tempted to keep the chain as dragon v0.8 and v0.9 use "CRAYEX" in the asset name, while v0.10 does (and presumably v0.11 will) use "HSN".

If not, totally willing to collapse this down to just the single `product`

Reply:
It is definitely meant to mimic those release assets. There are two packages per python version, for example:

- `dragon-0.10-py3.10.10-7edea3b6f.tar.gz`
- `dragon-0.10-py3.10.10-HSN-4e64b74bf.tar.gz`

In your sample test data, cray ex carries a sort of historical connotation since that was the name of the files before HSN replaced it. I think having CRAYEX does prove out that we'd chose the HSN version when we're supposed to, but that historical meaning makes it appear to be an important distinction, which it no longer is.

In the examples for py3.10 above, we just see that you either have a segment denoting HSN capability or you don't... so, i feel like we don't need to look for `FOO-\` or `CRAYEX-` as long as the hsn and non-hsn is covered.

Your first set of tests will all be negative for hsn.
- `["py3.9", "py3.10", "py3.11"], ["dragon-0.8", "dragon-0.9"], ["", "CRAYEX-"]`

The second set of tests exercise both hsn and non-hsn packages
- `["py3.9", "py3.10", "py3.11"], ["dragon-0.10", "dragon-0.11"], ["", "HSN-"]`

So, why have the first set of tests at all if they don't do anything the product with `""` already accomplishes.

Reply:
Oh! I can buy that!! I kept them around originally bc _technically_ there are assets that _would_ be found during the actual build process by the name `-CRAYEX` at version `v0.9`, but I can 100% get behind the argument that they are not actually _testing_ anything and are just adding noise to the tests. Will remove!